### DISCUSSION TITLE
Constraints do not work correctly when using ScriptFunctionInstance.Invoke

### DISCUSSION THREAD

**Discussion by dmsch** [sebastienros/jint#984] [Q&A] [answered]
I tried using ScriptFunctionInstance.Invoke to call Javascript callbacks from C# but get MemoryLimitExceededException or TimeoutException when called multiple times, although this, at least I think, should not be the case.

**Example which throws MemoryLimitExceededException :**
```C#
var engine = new Engine(options => { options.LimitMemory(1_000_000); });
            
engine.Execute("var callback = function(arg) {}");
var callback = engine.GetValue("callback") as ScriptFunctionInstance;
var testData = "Lorem Ipsum Dolor Sit Amet";
            
for (int i = 0; i < 100000; i++)
{
      callback.Invoke(testData); // Throws MemoryLimitExceededException after 2258 iterations.
}
```

**Example which throws TimeoutException:**
```C#
var engine = new Engine(options => { options.TimeoutInterval(TimeSpan.FromSeconds(5)); });
            
engine.Execute("var callback = function(arg) {}");
var callback = engine.GetValue("callback") as ScriptFunctionInstance;
var testData = "Lorem Ipsum Dolor Sit Amet";
            
for (int i = 0; i < 10; i++)
{
     callback.Invoke(testData); // Throws TimeoutException after 5 iterations
     Thread.Sleep(1000);
}
```

Is the ScriptFunctionInstance.Invoke method meant to be called externally, or would I be better off using Engine.Invoke instead?

**Comment by lahma** [comment, marked as answer]
Generally it's always better to route calls through Engine instance. Engine knows how to reset state between calls. 

You could even go one step further and set the `testData` as variable into engine and then call `engine.Evaluate("callback(testData)")`, that's something I find a natural way to interact.

**Reply by dmsch** [reply]
Thanks for your answer, I'll do that.
Just for information, using the Engine directly, with 
`engine.Invoke((ScriptFunctionInstance)callback, new[] { (JsValue)testData });`
also throws the same Exceptions.

**Reply by lahma** [reply]
If it was up to me I'd probably remove other methods than `Evaluate` and `Execute` for running JS code, other ways tend to get tricky 🤷🏻‍♂️

/cc @sebastienros

**Reply by lahma** [reply]
Maybe `Invoke` would make more sense if it would allow only string as first parameter (the target), dunno. @dmsch how does `engine.Invoke("callback", testData)` behave in this case?

**Reply by dmsch** [reply]
`engine.Invoke("callback", testData)` also throws these Exceptions.

Created an example that shows why I tried to call the callbacks directly: https://gist.github.com/dmsch/5a243ea5352bfb5cc1f3322481f95b0e

**Reply by lahma** [reply]
Now also `Invoke` should follow set constraints properly: https://github.com/sebastienros/jint/pull/986

**Reply by dmsch** [reply]
@lahma Thanks for the change. 
This library is top-notch.
